sdk/storage/azure-storage-datamovement/src/job_engine_task_generation.cpp

```diff
@@ -49,6 +51,108 @@ namespace Azure { namespace Storage { namespace _detail {
+      int64_t subtaskDownloadSize = 0;
+      int64_t totalDownloadSize = 0;
+
+      auto flushSubtask = [&]() {
+        auto description = Core::Json::_internal::json::array();
+        for (const auto& range : subtaskRanges)
+        {
+          description.push_back(range.Offset);
+          description.push_back(range.Length.Value());
+        }
+        subtaskRanges.clear();
+        if (!description.empty())
+        {
+          subtasksDescription.push_back(std::move(description));
+        }
+        totalDownloadSize += subtaskDownloadSize;
+        subtaskDownloadSize = 0;
+      };
+      std::function<void(const Azure::Core::Http::HttpRange&)> addRange;
+      addRange = [&](const Azure::Core::Http::HttpRange& range) {
+        constexpr int64_t MergeThreshold = 512;
+        AZURE_ASSERT(range.Length.Value() > 0);
+        int64_t distance = subtaskRanges.empty()
+            ? std::numeric_limits<int64_t>::max()
+            : range.Offset - (subtaskRanges.back().Offset + subtaskRanges.back().Length.Value());
+
+        if (range.Length.Value() > task.ChunkSize)
+        {
+          if (distance > 0 && distance <= MergeThreshold
+              && task.ChunkSize - subtaskDownloadSize > distance)
+          {
+            Azure::Core::Http::HttpRange newRange;
+            newRange.Offset = range.Offset - distance;
+            newRange.Length = distance + range.Length.Value();
+            addRange(newRange);
+          }
+          else if (task.ChunkSize - subtaskDownloadSize > 0)
+          {
+            Azure::Core::Http::HttpRange newRange;
+            newRange.Offset = range.Offset;
+            newRange.Length = task.ChunkSize - subtaskDownloadSize;
+            addRange(newRange);
+            Azure::Core::Http::HttpRange newRange2;
+            newRange2.Offset = range.Offset + newRange.Length.Value();
+            newRange2.Length = range.Length.Value() - newRange.Length.Value();
+            addRange(newRange2);
+          }
+          else
+          {
+            flushSubtask();
+            addRange(range);
+          }
+        }
+        else
+        { // range.Length.Value() <= task.ChunkSize
+          if (distance <= MergeThreshold
+              && subtaskDownloadSize + distance + range.Length.Value() <= task.ChunkSize)
+          {
+            subtaskRanges.back().Length.Value() += distance + range.Length.Value();
+            subtaskDownloadSize += distance + range.Length.Value();
+          }
+          else if (
+              distance > MergeThreshold
+              && subtaskDownloadSize + range.Length.Value() <= task.ChunkSize)
+          {
+            subtaskRanges.push_back(range);
+            subtaskDownloadSize += range.Length.Value();
+          }
+          else
+          {
+            flushSubtask();
+            addRange(range);
+          }
+        }
+      };
+
+      for (auto rangePage = pageBlobClient.GetPageRanges(); rangePage.HasPage();
```

Comment:
It could take long time or even timeout to GetpageRanges for a large page blob with large number of small ranges. Please change to get page ranges with spans or chunks.

Reply:
I checked the REST API, this API returns paged results since version `2020-10-02`, each page containing at most 10000 items. If this is the case, I don't think there's anything we need to do at client side.